### PR TITLE
Allow validators to register address to send report transactions - Implement handlers and keepers

### LINKED ISSUE #848: Allow validators to register address to send report transactions - Implement handlers and keepers

### PR DESCRIPTION
**Pull request by Benzbeeb**
Fixed #848

### CODE REVIEW

File: chain/cmd/bandoracled/main.go [hunk 1]
```diff
@@ -214,7 +214,7 @@ func handleRequest(requestID zoracle.RequestID) {
 		return
 	}
 	tx, err := bandClient.SendTransaction(
-		zoracle.NewMsgReportData(requestID, reports, sdk.ValAddress(bandClient.Sender())),
+		zoracle.NewMsgReportData(requestID, reports, sdk.ValAddress(bandClient.Sender()), sdk.AccAddress(bandClient.Sender())),
```
Comment (sorawit) [inline chain/cmd/bandoracled/main.go]: This the casting necessary?

File: chain/x/zoracle/client/cli/tx.go [hunk 2]
```diff
@@ -194,7 +194,7 @@ $ %s tx zoracle report 1 1:172.5 2:HELLOWORLD --from mykey
 				return dataset[i].ExternalDataID < dataset[j].ExternalDataID
 			})
 
-			msg := types.NewMsgReportData(requestID, dataset, sdk.ValAddress(cliCtx.GetFromAddress()))
+			msg := types.NewMsgReportData(requestID, dataset, sdk.ValAddress(cliCtx.GetFromAddress()), sdk.AccAddress(cliCtx.GetFromAddress()))
```
Comment (sorawit) [inline chain/x/zoracle/client/cli/tx.go]: ditto

File: chain/x/zoracle/internal/types/msgs_test.go [hunk 3]
```diff
@@ -133,17 +134,18 @@ func TestMsgReportDataValidation(t *testing.T) {
...
-		{true, NewMsgReportData(requestID, data, validator)},
-		{false, NewMsgReportData(-1, data, validator)},
-		{false, NewMsgReportData(requestID, []RawDataReportWithID{}, validator)},
-		{false, NewMsgReportData(requestID, nil, validator)},
-		{false, NewMsgReportData(requestID, data, failValidator)},
+		{true, NewMsgReportData(requestID, data, validator, reporter)},
+		{false, NewMsgReportData(-1, data, validator, reporter)},
+		{false, NewMsgReportData(requestID, []RawDataReportWithID{}, validator, reporter)},
+		{false, NewMsgReportData(requestID, nil, validator, reporter)},
+		{false, NewMsgReportData(requestID, data, failValidator, reporter)},
```
Comment (taobun) [inline chain/x/zoracle/internal/types/msgs_test.go:148]: Add invalid reporter too

File: chain/x/zoracle/handler.go [hunk 4]
```diff
@@ -26,6 +26,8 @@ func NewHandler(keeper Keeper) sdk.Handler {
 			return handleMsgRequestData(ctx, keeper, msg)
 		case MsgReportData:
 			return handleMsgReportData(ctx, keeper, msg)
+		case MsgAddOracleAddress:
+			return handleMsgAddOracleAddress(ctx, keeper, msg)
```
Comment (sorawit) [inline chain/x/zoracle/handler.go:30]: Where is MsgRemoveOracleAddress?

File: chain/x/zoracle/handler.go [hunk 5]
```diff
@@ -277,7 +279,39 @@ func handleMsgReportData(ctx sdk.Context, keeper Keeper, msg MsgReportData) sdk.
...
+	err := keeper.AddReporter(ctx, msg.Validator, msg.Reporter)
+	if err != nil {
+		return err.Result()
+	}
+
+	// Emit report event
+	ctx.EventManager().EmitEvents(sdk.Events{
+		sdk.NewEvent(
+			types.EventTypeAddOracleAddress,
+			sdk.NewAttribute(types.AttributeKeyValidator, msg.Validator.String()),
```
Comment (sorawit) [inline chain/x/zoracle/handler.go:301]: Add reporter here too

File: chain/x/zoracle/handler.go [hunk 6]
```diff
@@ -277,7 +279,39 @@ func handleMsgReportData(ctx sdk.Context, keeper Keeper, msg MsgReportData) sdk.
...
+	err := keeper.RemoveReporter(ctx, msg.Validator, msg.Reporter)
+	if err != nil {
+		return err.Result()
+	}
+
+	// Emit report event
+	ctx.EventManager().EmitEvents(sdk.Events{
+		sdk.NewEvent(
+			types.EventTypeRemoveOracleAddress,
+			sdk.NewAttribute(types.AttributeKeyValidator, msg.Validator.String()),
```
Comment (sorawit) [inline chain/x/zoracle/handler.go:318]: ditto

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/x/zoracle/alias.go [hunk 7]
```diff
@@ -66,13 +66,15 @@ var (
...
-	MsgEditOracleScript   = types.MsgEditOracleScript
+	Keeper                  = keeper.Keeper
+	MsgRequestData          = types.MsgRequestData
+	MsgReportData           = types.MsgReportData
+	MsgCreateDataSource     = types.MsgCreateDataSource
+	MsgEditDataSource       = types.MsgEditDataSource
+	MsgCreateOracleScript   = types.MsgCreateOracleScript
+	MsgEditOracleScript     = types.MsgEditOracleScript
+	MsgAddOracleAddress     = types.MsgAddOracleAddress
+	MsgRemoveOracleAdderess = types.MsgRemoveOracleAdderess
```
Comment (taobun) [inline chain/x/zoracle/alias.go:79]: Add NewMsg... in var

File: chain/x/zoracle/client/cli/tx.go (hunk 2 above)
Comment (taobun) [reply chain/x/zoracle/client/cli/tx.go]: Add // TODO: Receive validator address as flag

File: chain/x/zoracle/internal/keeper/report.go [hunk 8]
```diff
@@ -42,10 +43,22 @@ func (k Keeper) AddReport(
...
 	}
+
+	if k.CheckReporter(ctx, validator, reporter) {
+		agent := sdk.ValAddress(reporter)
+		for _, validValidator := range request.RequestedValidators {
+			if agent.Equals(validValidator) {
+				found = true
+				break
+			}
+		}
```
Comment (taobun) [inline chain/x/zoracle/internal/keeper/report.go]: Why we need this code?

File: chain/x/zoracle/internal/keeper/report.go [hunk 9]
```diff
@@ -42,10 +43,22 @@ func (k Keeper) AddReport(
 			break
 		}
 	}
+
```
Comment (taobun) [inline chain/x/zoracle/internal/keeper/report.go:54]: If found is false, we must return err with validator not in list?

File: chain/x/zoracle/internal/keeper/report.go (hunk 8 above)
Comment (taobun) [reply chain/x/zoracle/internal/keeper/report.go]: And if this reporter is not an illegal reporter of the validator.  It should return another error

File: chain/x/zoracle/internal/keeper/report_test.go [hunk 10]
```diff
@@ -62,7 +62,7 @@ func TestAddReportSuccess(t *testing.T) {
 	err = keeper.AddReport(ctx, 1, []types.RawDataReportWithID{
 		types.NewRawDataReportWithID(2, 0, []byte("data1/2")),
 		types.NewRawDataReportWithID(10, 2, []byte("data2/2")),
-	}, sdk.ValAddress([]byte("validator2")))
+	}, sdk.ValAddress([]byte("validator2")), sdk.AccAddress([]byte("validator1")))
```
Comment (taobun) [inline chain/x/zoracle/internal/keeper/report_test.go]: Why this test passed?

File: chain/x/zoracle/internal/keeper/report_test.go [hunk 11]
```diff
@@ -230,13 +230,13 @@ func TestAddReportReportSizeExceedMaxRawDataReportSize(t *testing.T) {
...
 		types.NewRawDataReportWithID(2, 0, []byte("short report")),
-	}, sdk.ValAddress([]byte("validator1")))
+	}, sdk.ValAddress([]byte("validator1")), sdk.AccAddress([]byte("validator1")))
 	require.Nil(t, err)
 
 	// Size of "a report that obviously longer than 20 bytes" is 44 bytes.
 	err = keeper.AddReport(ctx, 1, []types.RawDataReportWithID{
 		types.NewRawDataReportWithID(2, 0, []byte("a report that obviously longer than 20 bytes")),
-	}, sdk.ValAddress([]byte("validator2")))
+	}, sdk.ValAddress([]byte("validator2")), sdk.AccAddress([]byte("validator2")))
```
Comment (taobun) [inline chain/x/zoracle/internal/keeper/report_test.go:239]: Add new unit test to test invalid reporter

**Review by taobun — CHANGES_REQUESTED** [review]

File: chain/x/zoracle/internal/keeper/report.go [hunk 12]
```diff
@@ -35,17 +36,27 @@ func (k Keeper) AddReport(
...
+			"AddReport: Reporter (%s) is unauthorized permission to be reporter.",
+			reporter.String(),
+		)
+	}
+
 	found := false
+	agent := sdk.ValAddress(reporter)
 	for _, validValidator := range request.RequestedValidators {
-		if validator.Equals(validValidator) {
+		if agent.Equals(validValidator) {
```
Comment (sorawit) [inline chain/x/zoracle/internal/keeper/report.go]: Is this correct?

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/x/zoracle/internal/keeper/report.go [hunk 13]
```diff
@@ -35,17 +36,27 @@ func (k Keeper) AddReport(
 		)
 	}
 
+	if !k.CheckReporter(ctx, validator, reporter) {
+		return types.ErrUnauthorizedPermission(
+			"AddReport: Reporter (%s) is unauthorized permission to be reporter.",
```
Comment (sorawit) [inline chain/x/zoracle/internal/keeper/report.go]: %s is not an authorized reporter of %s

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/x/zoracle/internal/keeper/report.go [hunk 14]
```diff
@@ -35,17 +36,27 @@ func (k Keeper) AddReport(
...
+		if agent.Equals(validValidator) {
 			found = true
 			break
 		}
 	}
+
 	if !found {
 		return types.ErrUnauthorizedPermission(
-			"AddReport: Reporter (%s) is not on the reporter list.",
+			"AddReport: Reporter (%s) and (%s) is not on the reporter list.",
```
Comment (sorawit) [inline chain/x/zoracle/internal/keeper/report.go]: Validator (%s) is not on the reporter list.

**Review by sorawit — CHANGES_REQUESTED** [review]

**Review by sorawit — APPROVED** [review]
Looks good

**Review by taobun — APPROVED** [review]

**Pull request merged**